tree/dataframe/inc/ROOT/RDF/RMergeableValue.hxx

```diff
@@ -0,0 +1,458 @@
+know how to merge with other results of the same type. In a distributed
+environment it is often needed to have a merging mechanism for partial results
+coming from the execution of an analysis on different chunks of the same dataset
+that has happened on different executors. In order to achieve this,
+RMergeableValue stores the result of the RDataFrame action and has a `Merge`
+method to allow the aggregation of information coming from another similar
+result into the current.
+
+A mergeable value can be retrieved from an RResultPtr through the
+`GetMergeableValue` free function and a sequence of mergeables can be merged
+together with the helper function `MergeValues`. All the classes and functions
+involved are inside the `ROOT::Detail::RDF` namespace.
+
+In a nutshell:
+~~~{.cpp}
+using namespace ROOT::Detail::RDF;
+ROOT::RDataFrame d("myTree", "file_*.root");
+auto h1 = d.Histo1D("Branch_A");
+auto h2 = d.Histo1D("Branch_A");
+
+// Retrieve mergeables from the `RResultPtr`s
+auto mergeableh1 = GetMergeableValue(h1);
+auto mergeableh2 = GetMergeableValue(h2);
+
+// Merge the values and get another mergeable back
+auto mergedptr = MergeValues(std::move(mergeableh1), std::move(mergeableh2));
+
+// Retrieve the merged TH1D object
+const auto &mergedhisto = mergedptr->GetValue();
+~~~
+
+Though this snippet can run on a single thread of a single machine, it is
+straightforward to generalize it to a distributed case, e.g. where `mergeableh1`
+and `mergeableh2` are created on separate machines and sent to a `reduce`
+process where the `MergeValues` function is called. The final user would then
+just be given the final merged result coming from `mergedptr->GetValue`.
+
+RMergeableValue is the base class for all the different specializations that may
+be needed according to the peculiarities of the result types. The following
+subclasses are currently available:
+
+| RMergeableValue Specialization | Corresponding RDataFrame Action(s)         |
+|--------------------------------|--------------------------------------------|
+| RMergeableFill                 | Graph, Histo{1,2,3}D, Profile{1,2}D, Stats |
+| RMergeableGenericAction        | Count, Max, Min, Sum                       |
+| RMergeableMean                 | Mean                                       |
+| RMergeableStdDev               | StdDev                                     |
+*/
+template <typename T>
+class RMergeableValue : public RMergeableValueBase {
+   // Friend function declaration
+   template <typename T1, typename... Ts>
+   friend std::unique_ptr<RMergeableValue<T1>> MergeValues(std::unique_ptr<RMergeableValue<T1>> OutputMergeable,
+                                                           std::unique_ptr<RMergeableValue<Ts>>... InputMergeables);
+   template <typename T1, typename... Ts>
+   friend void MergeValues(RMergeableValue<T1> &OutputMergeable, const RMergeableValue<Ts> &... InputMergeables);
```

Comment:
can these `T1`s be `T`s instead?

Reply:
Seems like they can't:
* Writing `T` in the template declaration shadows the `T` of the class template
* Writing `template<typename ... Ts> friend void MergeValues<T>` gives error about not being able to make partial template instantiation

Reply:
alright :man_shrugging: